zipline/gens/sim_engine.pyx

```diff
@@ -59,28 +64,50 @@ cdef class MinuteSimulationClock:
 
     @cython.boundscheck(False)
     @cython.wraparound(False)
-    cdef dict calc_minutes_by_day(self):
-        cdef dict minutes_by_day
-        cdef int day_idx
-        cdef object day
+    cdef dict calc_minutes_by_session(self):
+        cdef dict minutes_by_session
+        cdef int session_idx
+        cdef object session
 
-        minutes_by_day = {}
-        for day_idx, day in enumerate(self.trading_days):
-            minutes_by_day[day] = pd.to_datetime(
-                self.market_minutes(day_idx), utc=True, box=True)
-        return minutes_by_day
+        minutes_by_session = {}
+        for session_idx, session in enumerate(self.sessions):
+            minutes_by_session[session] = pd.to_datetime(
+                self.market_minutes(session_idx), utc=True, box=True)
+        return minutes_by_session
 
     def __iter__(self):
         minute_emission = self.minute_emission
 
-        for day in self.trading_days:
-            yield day, DAY_START
+        for idx, session in enumerate(self.sessions):
+            yield session, SESSION_START
+
+            bts_minute = self.before_trading_start_minutes[idx]
+            regular_minutes = self.minutes_by_session[session]
+
+            # we have to search anew every session, because there is no
+            # guarantee that any two session start on the same minute
```

Comment:
Sure.  A trading calendar can have late opens, which means that for two trading sessions, the index of the `bts_minute` in the sessions' minutes can be different.